### DISCUSSION TITLE
Generic interface so any model (regardless of implementation) can be used as RM / LM?

### DISCUSSION THREAD

**Discussion by vanguardapps** [stanfordnlp/dspy#444] [Q&A] [answered]
Hi. I do apologize if this is in the docs somewhere, but so far all I can see if the experimental feature in the works for supporting a local Hugging Face non-TGI-setup. What I am looking for is a generic class-based interface that allows us to attach any model we want, regardless of where it is hosted, what API interface it uses, etc (we have to write that part) to DSPy in a simple, general abstraction.

This is something that I think would be helpful, because all models perform essentially the same thing: accept input, give output. Please let me know if there is something in here that can be used this way.

For clarity, what I'm imagining is just something that takes a generic `**kwargs` and returns a generic `**kwargs`, perhaps with some binding restrictions on naming, that can be used with a subset of DSPy functionality. Perhaps for training / finetuning models (and especially for compiling), you would need to dive deeper and integrate further, providing more details and conforming to more restrictions (similar to how TGI works).

I guess I just don't see a reason not to have some kind of open-ended thing where anyone can integrate any model they want, limited only by how replete their integration is to the extent that certain DSPy features might not be usable (for instance, if the integration only includes inference on a frozen model).

I don't see this as an issue per se, more of a question reaching out to the contributers, and was hoping for just a little feedback to see if my thinking is way off. Thank you for your time reading.

**Comment by vanguardapps** [comment, marked as answer]
Okay, so I am answering my own discussion question. My hope is that someone finds this with the same question and will find this link useful: https://dspy-docs.vercel.app/docs/deep-dive/language_model_clients/custom-lm-client

This is provided! (And it's funny how similar it actually is to what I described up there, with the `**kwargs` going both ways and all.)

Great work. I'm happy to see this.

**Comment by kinthaiofficial** [comment]
The generic model interface question is central to DSPy's portability — and the current state is better than it was but still has rough edges.

The core abstraction that works:

```python
import dspy

# DSPy's LM abstraction wraps any model that supports chat/completion
lm = dspy.LM("openai/gpt-4o")           # OpenAI
lm = dspy.LM("anthropic/claude-3-5-sonnet")  # Anthropic
lm = dspy.LM("ollama/llama3", api_base="http://localhost:11434")  # Local
lm = dspy.LM("huggingface/mistralai/Mistral-7B-v0.1")  # HuggingFace

dspy.configure(lm=lm)
```

**Where it breaks down**:

1. **Chat template differences**: Some models (especially smaller fine-tunes) need specific system/user/assistant formatting that DSPy's templates may not match exactly. The `chat_template` argument helps but requires knowing the model's expected format.

2. **Tokenizer access for few-shot selection**: DSPy's optimizers need to know token counts to fit examples in context. For non-OpenAI models, the token counting is approximate.

3. **Logprobs**: Some optimizers use logprob-based scoring for optimization. HuggingFace supports this natively; API providers vary.

```python
# Workaround for models with specific chat templates
class CustomModel(dspy.LM):
    def __init__(self, model_path):
        self.tokenizer = AutoTokenizer.from_pretrained(model_path)
        self.pipeline = pipeline("text-generation", model=model_path)
    
    def __call__(self, prompt, **kwargs):
        formatted = self.tokenizer.apply_chat_template(
            [{"role": "user", "content": prompt}],
            tokenize=False
        )
        return self.pipeline(formatted, max_new_tokens=512)[0]["generated_text"]
```

Is the RFC targeting a specific failure mode — model that works with LangChain but fails with DSPy? That'd help narrow the interface gap.
